rewrite-java/src/test/java/org/openrewrite/java/OpenRewriteArchitectureTest.java

```diff
@@ -0,0 +1,81 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java;
+
+import com.tngtech.archunit.core.domain.JavaClasses;
+import com.tngtech.archunit.core.domain.JavaPackage;
+import com.tngtech.archunit.core.importer.ClassFileImporter;
+import com.tngtech.archunit.core.importer.ImportOption;
+import com.tngtech.archunit.lang.*;
+import org.junit.jupiter.api.Test;
+import org.openrewrite.internal.lang.NonNullApi;
+
+import static com.tngtech.archunit.base.DescribedPredicate.doNot;
+import static com.tngtech.archunit.core.domain.properties.HasName.Predicates.name;
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.*;
+
+public class OpenRewriteArchitectureTest {
+    JavaClasses importedClasses = new ClassFileImporter()
+      .withImportOption(new ImportOption.DoNotIncludeTests())
+      .importPackages("org.openrewrite.java");
+
+    @Test
+    void packageInfo() {
+        all(packages("org.openrewrite.java"))
+          .that(doNot(name("org.openrewrite.java.internal.grammar")))
+          .should(containAnAnnotatedPackageInfo())
+          .check(importedClasses);
+    }
+
+    @Test
+    void nullness() {
+        noCodeUnits()
+          .should().beAnnotatedWith("org.jetbrains.annotations.NotNull")
+          .orShould().beAnnotatedWith("org.jetbrains.annotations.Nullable")
+          .check(importedClasses);
+    }
+
+    private static ClassesTransformer<JavaPackage> packages(String basePackage) {
+        return new AbstractClassesTransformer<>("packages") {
+            @Override
+            public Iterable<JavaPackage> doTransform(JavaClasses classes) {
+                return classes.getPackage(basePackage).getSubpackagesInTree();
+            }
+        };
+    }
+
+    private static ArchCondition<JavaPackage> containAnAnnotatedPackageInfo() {
+        return new ArchCondition<JavaPackage>("contain a package-info") {
+            @Override
+            public void check(JavaPackage javaPackage, ConditionEvents events) {
+                if (javaPackage.tryGetPackageInfo().isEmpty()) {
+                    String message = String.format("Package '%s' does not contain a package-info", javaPackage.getName());
+                    events.add(SimpleConditionEvent.violated(javaPackage, message));
+                }
+            }
+        }.and(new ArchCondition<>("be annotated with @NonNullApi") {
+            @Override
+            public void check(JavaPackage javaPackage, ConditionEvents events) {
+                if (javaPackage.tryGetPackageInfo().isPresent()) {
+                    if (!javaPackage.getPackageInfo().isAnnotatedWith(NonNullApi.class)) {
+                        String message = String.format("Package '%s' is not annotated as @NonNullApi", javaPackage.getName());
+                        events.add(SimpleConditionEvent.violated(javaPackage, message));
+                    }
+                }
```

Comment:
`isPresent()` within an if statement always looks a bit odd to me, but maybe that's personal.
Had you considered something like
```suggestion
                javaPackage.tryGetPackageInfo()
                        .filter(packageInfo -> !packageInfo.isAnnotatedWith(NonNullApi.class))
                        .ifPresent(packageInfo -> events.add(SimpleConditionEvent.violated(javaPackage,
                                String.format("Package '%s' is not annotated as @NonNullApi", javaPackage.getName()))));
```